File: 1 Kyu/1 kyu - Square sums.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>

using namespace std;

struct vertex {
  int id, visit;
  vector<vertex*> edges;
};

class Timer {
    //using time = chrono::steady_clock;
    private :
        chrono::steady_clock::time_point alpha, end;
        chrono::duration<double> elapsed;
        uint64_t cycle;

        void update () {
            end = chrono::steady_clock::now ();
            elapsed = end - alpha;
        }

    public :

        Timer() {
            alpha = chrono::steady_clock::now ();
            cycle = 0;
        }
        void start () { alpha = chrono::steady_clock::now ();}
        void stop () { update();}
        void get_duration () {
            std::cout << "\nDuration "
                      <<fixed<< elapsed.count()
                      << " ms" << std::endl;
        }
        bool running (double total) {
            update();
            cycle++;
            if (elapsed.count() < total) return true;
            cout << "cycles :: " ;

            string num = to_string (cycle);
            int size = num.size(), index = 0;

            while (size-->0) {
                cout << num[index++];
                if (size % 3 == 0) cout << ' ';
            }
            cout << endl;
            return false;
        }
};
class Display {
    public :

        static void graph (vertex **src) {
            for (int i = 1; src[i] ; i++) {

            }
            cout << endl;
        }
        static void vect (vector<int> src) {

            for (auto &it : src)
                cout << it << " ";

            cout << endl;
        }
};

class graph {
      private :
          vector<int> sqr;
          vector<vertex> vmap;

      public :
          int size;

          graph (int n) {
              size = n;

              for (int i = 2; i * i < (size * 2); i++) {
                  sqr.push_back (i * i);
              }

              for (int i = 0; i < size + 1 ; i++) {
                  vmap.push_back ({i, false, vector<vertex *>()});
              }

              for (int i = 1; i < size + 1; i++) {
                  update (&vmap[i]);
              }

          }
          void update (vertex *v) {

              v->edges.clear();
              for (auto &sq : sqr) {
                  int val = sq - v->id;

                  if (val > 0 && val <= size && val != v->id && !vmap[val].visit) {
                        v->edges.push_back (&vmap[val]);
                  }
              }
          }

          vertex *operator [] (int index) { return &vmap[index]; }
};

int getstart (graph &curr) {
  size_t minv = 99, start;

  for (int i = 1; i < curr.size + 1; i++) {
      if (curr[i]->edges.size() <= minv) {
          minv = curr[i]->edges.size();
          start = i;
      }
  }
  return start;
}
bool dfs (size_t index, graph &curr, vector<int> &seq) {

    if (index == seq.size() - 1) return true;

    vertex *u = curr[seq[index]];

    for (auto edge : u->edges) {
        curr.update (edge);
    }

    sort (u->edges.begin(), u->edges.end(), [] (const vertex *a, const vertex *b) {
        int aa = a->edges.size(), bb = b->edges.size();
        return (aa != bb) ? aa < bb : a->id < b->id;
    });

    for (auto edge : u->edges) {
        seq[index + 1] = edge->id;
        edge->visit = true;

        if (dfs (index + 1, curr, seq)) return true;

        seq[index + 1] = 0;
        edge->visit = false;
    }

    return false;
}
vector<int> square_sums_row (int size) {

    graph curr (size);
    vector<int> seq (size);

    seq[0] = getstart (curr);
    curr[seq[0]]->visit = true;
    // cout << seq[0] << ' ';
    return dfs (0, curr, seq) == true ? seq : vector<int>{};
}

int main () {

    vector<int> test = {15, 23, 50, 56, 128, 256, 512, 980};
    vector<int> vec;

    for (auto N : test) {
        Timer clock;
        auto res = square_sums_row(N);

        // std::cout << res.size() << ' ';
        // Display::vect (res);

        clock.stop();
        clock.get_duration();
    }

}

File: 1 Kyu/1 kyu - Square sums.py
import time
import sys

sys.setrecursionlimit(1100)
start = time.time()

class vertex :
    def __init__ (self, id) :
        self.id = id
        self.visit = False
        self.edges = []

class Graph :
    def __init__ (self, N) :
        self.squares = [i * i for i in range(2, int(N ** 0.5) + 1)]
        self.vmap = [vertex(i) for i in range(0,N+1)]

        for i in range(1, N + 1) : self.update(self.vmap[i])

    def update (self, node) :
        node.edges = []

        for sq in self.squares :
            j = sq - node.id
            if j > 0 and j < len(self.vmap) and j != node.id and self.vmap[j].visit == False :
                node.edges.append(j)

def getstart (vmap) :
    start, minv = 0, 99

    for i in range(1, len(vmap)) :
        if len(vmap[i].edges) <= minv :
            minv = len(vmap[i].edges)
            start = i

    return start

def dfs (index, seq, curr) :

    if index == len(seq) - 1 : return True

    graph = curr.vmap
    node = graph[seq[index]]

    for edge in node.edges :
        curr.update(graph[edge])

    node.edges.sort(key=lambda x: len(graph[x].edges))

    for edge in node.edges :
        if graph[edge].visit == False :
            graph[edge].visit = True
            seq[index + 1] = edge

            if dfs(index + 1, seq, curr)  : return True

            seq[index + 1] = 0
            graph[edge].visit = False

    return False

def square_sums (N) :

    curr = Graph(N)
    seq = [0] * N

    graph = curr.vmap
    seq[0] = getstart(graph)
    graph[seq[0]].visit = True

    return seq if dfs(0, seq, curr) == True else False


for num in [5, 15, 23, 50, 56, 128, 512, 993, 1000] :
    res = square_sums(num)


end = time.time()
print( "elapsed : ", end - start)
